src/Storages/MergeTree/MutateTask.cpp

```diff
@@ -2359,6 +2328,61 @@ static bool canSkipMutationCommandForPart(const MergeTreeDataPartPtr & part, con
     return false;
 }
 
+namespace
+{
+/// Calculate the set of indices which should be recalculated and dropped during mutation.
+/// Also wraps the input stream into additional expression stream
+void updateIndicesToRecalculateAndDrop(std::shared_ptr<MutationContext> & ctx)
+{
+    const MergeTreeDataPartPtr & source_part = ctx->source_part;
+    QueryPipelineBuilder & builder = ctx->mutating_pipeline_builder;
+    const StorageMetadataPtr & metadata_snapshot = ctx->metadata_snapshot;
+
+    /// Checks if columns used in skipping indexes modified.
+    const auto & index_factory = MergeTreeIndexFactory::instance();
+    ASTPtr indices_recalc_expr_list = std::make_shared<ASTExpressionList>();
+    const auto & indices = metadata_snapshot->getSecondaryIndices();
+    bool is_full_part_storage = isFullPartStorage(source_part->getDataPartStorage());
+
+    for (const auto & index : indices)
+    {
+        if (ctx->indices_to_drop_names.contains(index.name))
+        {
+            ctx->indices_to_drop.insert(index_factory.get(index));
+            continue;
+        }
+
+        bool need_recalculate
+            = ctx->materialized_indices.contains(index.name) || (!is_full_part_storage && source_part->hasSecondaryIndex(index.name));
+
+        if (need_recalculate)
+        {
+            if (ctx->indices_to_recalc.insert(index_factory.get(index)).second)
+            {
+                ASTPtr expr_list = index.expression_list_ast->clone();
+                for (const auto & expr : expr_list->children)
+                    indices_recalc_expr_list->children.push_back(expr->clone());
```

Comment:
[minor] do we need to clone the expr from the cloned list? can we avoid allocation and just move?